### PR TITLE
Fix setoid_rewrite on Equiv : Type -> Type -> Type

### LINKED ISSUE #6278: setoid_rewrite fails on Equiv : Type -> Type -> Type
The following setoid_rewrite fails with a universe error.
```coq
Require Import Classes.CMorphisms Classes.CRelationClasses.
Require Setoids.Setoid.

Axiom Equiv : Type -> Type -> Type.

Instance equivalence_equiv : CRelationClasses.Equivalence Equiv.
Admitted.

Instance prod_equiv : CMorphisms.Proper (Equiv ==> Equiv ==> Equiv) prod.
Admitted.

Lemma foo {A B C:Type} (X: Equiv A B) : Equiv (prod C A) (prod C B).
Proof.
  setoid_rewrite X.
```
The error is:
```
The 3rd term has type
 "crelation@{Top.57 Top.58} Type@{max(Top.22, Coq.Init.Datatypes.24)}"
which should be coercible to "crelation@{Top.83 Top.84} Type@{Top.1}".
```
The problem seems to be the type class search as the `eapply` works whereas nor `typeclasses_eauto` nor `eauto with typeclass_instances` does.

Changing `Equiv` to `Type -> Type -> Prop` does not solve the problem (but the error message is different) but changing it to `myType -> myType -> myType` with `myType := Type` monomorphic does.

This bug may also be related to https://github.com/coq/coq/issues/3254 but I'm not able to affirm it.

---

My goal is to use `setoid_rewrite` to rewrite equivalences in the HoTT library, as in https://github.com/HoTT/HoTT/issues/488. Do you think the implementation feasible?

EDIT now the error is "nothing to rewrite"

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #6278

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Witness the cost of universes! https://gitlab.com/coq/coq/-/jobs/1986276361

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
What I still don't understand about this is before changing `setoid_rewrite` to `rewrite` fixed the test. Why would these be different?

**Comment by Alizter** [comment]
@SkySkimmer Can you rebase?

**Comment by ppedrot** [comment]
> Witness the cost of universes!

What were we supposed to see? It don't observe anything.

**Comment by Alizter** [comment]
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬───────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │    mem faults     │
│                             │                         │                                       │                                       │                         │                   │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD    PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼───────────────────┤
│                    coq-core │  101.09   102.33  -1.21 │   374705553208    373718374888   0.26 │   433429094612    433436320884  -0.00 │  272708   273880  -0.43 │   1    0      nan │
│                    coq-hott │  157.20   157.79  -0.37 │   682901682203    682868959576   0.00 │  1074023178081   1074058433050  -0.00 │  659928   656092   0.58 │   0    0      nan │
│       coq-engine-bench-lite │  180.37   180.99  -0.34 │   757782204468    760851171769  -0.40 │  1484339118338   1484576055710  -0.02 │ 1172576  1172468   0.01 │   0    0      nan │
│        coq-mathcomp-algebra │   64.64    64.86  -0.34 │   282837979587    283423455836  -0.21 │   395669277178    395688375142  -0.00 │  558460   556660   0.32 │   0    0      nan │
│       coq-mathcomp-fingroup │   23.96    24.04  -0.33 │   104811099002    105148836720  -0.32 │   155011693914    154977651716   0.02 │  477476   477476   0.00 │   0    0      nan │
│                   coq-color │  260.43   260.86  -0.16 │  1130360634793   1131935915019  -0.14 │  1539163676035   1539055412035   0.01 │ 1147792  1147412   0.03 │   0    1  -100.00 │
│       coq-mathcomp-solvable │   89.49    89.61  -0.13 │   391350175193    392577627694  -0.31 │   609131484473    609182370564  -0.01 │  669672   670140  -0.07 │   0    0      nan │
│         coq-metacoq-erasure │   64.81    64.88  -0.11 │   278793068867    279148081647  -0.13 │   423443679215    423291813971   0.04 │ 1070100  1069932   0.02 │   0    0      nan │
│                coq-coqprime │  185.54   185.74  -0.11 │   811092443412    812797107171  -0.21 │  1554491743008   1554549768127  -0.00 │  906160   905960   0.02 │   0    2  -100.00 │
│            coq-fiat-parsers │  340.20   340.46  -0.08 │  1470965767298   1470061019031   0.06 │  2435880749982   2435837504335   0.00 │ 2782068  2782200  -0.00 │   0    1  -100.00 │
│        coq-metacoq-template │   75.26    75.29  -0.04 │   321262134229    322207047924  -0.29 │   489772512923    489805822016  -0.01 │  697196   698624  -0.20 │   0    0      nan │
│            coq-math-classes │   95.92    95.94  -0.02 │   416362694873    415830302399   0.13 │   587798940299    587735524344   0.01 │  495564   495988  -0.09 │   0    0      nan │
│               coq-fourcolor │ 1522.20  1522.28  -0.01 │  6657307698371   6657017054944   0.00 │ 12081249073874  12081401856740  -0.00 │  725572   723952   0.22 │   0    0      nan │
│      coq-mathcomp-odd-order │  572.82   572.76   0.01 │  2510800984896   2510616910774   0.01 │  4279765878100   4279733805996   0.00 │  932376   932228   0.02 │   0    0      nan │
│             coq-fiat-crypto │ 2956.26  2955.61   0.02 │ 12900524162368  12899335690570   0.01 │ 23776512407422  23773799154585   0.01 │ 3794472  3794308   0.00 │  87   96    -9.38 │
│                  coq-flocq3 │   78.85    78.83   0.03 │   343749236481    342963145820   0.23 │   454441515462    454444876718  -0.00 │  993024   989524   0.35 │  85    0      nan │
│               coq-perennial │ 4095.56  4094.43   0.03 │ 17910127258750  17902613439402   0.04 │ 29881031990541  29880866904762   0.00 │ 2602620  2605200  -0.10 │   0    0      nan │
│               coq-fiat-core │   54.85    54.83   0.04 │   230052849903    229208231480   0.37 │   322620674457    322657113538  -0.01 │  476188   475776   0.09 │   1    3   -66.67 │
│     coq-metacoq-safechecker │  190.19   190.12   0.04 │   830450793223    829695097304   0.09 │  1125273012728   1125519017712  -0.02 │ 1179136  1182648  -0.30 │  12    0      nan │
│ coq-rewriter-perf-SuperFast │  944.76   944.24   0.06 │  4128313493117   4126957988967   0.03 │  7153493298045   7153881337503  -0.01 │ 1143224  1143616  -0.03 │   5    0      nan │
│              coq-coquelicot │   35.28    35.26   0.06 │   152259767253    151711916274   0.36 │   201749624856    201688114924   0.03 │  745304   746996  -0.23 │  87    0      nan │
│                    coq-corn │  790.91   790.18   0.09 │  3453287729122   3450774096066   0.07 │  5364365743290   5364419349344  -0.00 │  867836   870088  -0.26 │   0    0      nan │
│                     coq-vst │ 1088.03  1086.98   0.10 │  4751566907054   4746639646255   0.10 │  7796052503532   7796070092705  -0.00 │ 2115188  2115024   0.01 │   2    8   -75.00 │
│                coq-rewriter │  379.74   379.35   0.10 │  1656701039655   1654787504094   0.12 │  2719940814377   2719946620136  -0.00 │ 1058052  1058096  -0.00 │   7   11   -36.36 │
│                  coq-stdlib │  413.39   412.96   0.10 │  1462764414496   1455950948541   0.47 │  1405634406705   1405647633037  -0.00 │  598060   596372   0.28 │   0    0      nan │
│          coq-mathcomp-field │  115.06   114.92   0.12 │   504398926662    503489205104   0.18 │   838567720600    838586680186  -0.00 │  651660   649824   0.28 │   0    0      nan │
│                 coq-unimath │ 3784.87  3780.26   0.12 │ 16571079658127  16549326309801   0.13 │ 33164545791610  33164842960167  -0.00 │ 4125816  4126008  -0.00 │   1    0      nan │
│  coq-performance-tests-lite │  800.13   799.15   0.12 │  3481474785673   3477007935510   0.13 │  5918019651216   5917360879218   0.01 │ 2322396  2320544   0.08 │   2    0      nan │
│           coq-iris-examples │  434.27   433.73   0.12 │  1895837548629   1893455578164   0.13 │  2917462078531   2917653721863  -0.01 │ 1168312  1168284   0.00 │   1    0      nan │
│              coq-verdi-raft │  581.00   580.26   0.13 │  2537338113730   2534688440525   0.10 │  3940583931129   3940700143575  -0.00 │ 1220816  1220692   0.01 │   0    0      nan │
│      coq-mathcomp-character │   78.28    78.17   0.14 │   342970625628    342479008398   0.14 │   525225696848    525243544918  -0.00 │  724964   724128   0.12 │   0    0      nan │
│    coq-metacoq-translations │   20.20    20.17   0.15 │    86891860190     86726952437   0.19 │   152951007650    152951684355  -0.00 │  929824   929528   0.03 │   0    0      nan │
│                  coq-geocoq │  715.01   713.70   0.18 │  3121383313119   3113743631063   0.25 │  4943316734154   4943297430166   0.00 │ 1073732  1073724   0.00 │   9    0      nan │
│                coq-compcert │  287.31   286.54   0.27 │  1248335418871   1245843945099   0.20 │  1877123775989   1876640281226   0.03 │ 1098720  1097480   0.11 │   0    3  -100.00 │
│                 coq-coqutil │   35.64    35.52   0.34 │   153681526329    153150857455   0.35 │   207697919462    207718114129  -0.01 │  601436   601224   0.04 │   4    0      nan │
│                coq-bedrock2 │  193.79   193.08   0.37 │   845605856866    843380701780   0.26 │  1565327805132   1565364820262  -0.00 │ 2049460  2049332   0.01 │  17   12    41.67 │
│                   coq-verdi │   49.64    49.43   0.42 │   214760009412    214533514316   0.11 │   321950692112    322000737904  -0.02 │  832984   832104   0.11 │   0    0      nan │
│           coq-metacoq-pcuic │  497.83   495.29   0.51 │  2178867348339   2166249306471   0.58 │  3111637689308   3111355768482   0.01 │ 1119484  1119560  -0.01 │   2    0      nan │
│      coq-mathcomp-ssreflect │   26.41    26.24   0.65 │   115187505643    114690974438   0.43 │   146849475721    146936550782  -0.06 │  533932   531728   0.41 │   0    0      nan │
│                 coq-bignums │   28.72    28.49   0.81 │   125055655844    124650704005   0.32 │   172203299987    172211469618  -0.00 │  464756   468812  -0.87 │   0    0      nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴───────────────────┘
```

**Comment by SkySkimmer** [comment]
>What were we supposed to see? It don't observe anything.

Then universes cost nothing today.

File: plugins/ltac/rewrite.ml [hunk 1]
```diff
@@ -739,19 +736,22 @@ let unify_eqn (car, rel, prf, c1, c2, holes, sort) l2r flags env (sigma, cstrs)
...
-    let () = if not (convertible env evd ty2 ty1) then raise Reduction.NotConvertible in
-    let rew_evars = evd, cstrs in
-    let rew_prf = RewPrf (rel, prf) in
-    let rew = { rew_evars; rew_prf; rew_car; rew_from = c1; rew_to = c2; } in
-    let rew = if l2r then rew else symmetry env sort rew in
-    Some rew
+    let ty1 = Retyping.get_type_of env sigma c1 in
+    let ty2 = Retyping.get_type_of env sigma c2 in
+    begin match Reductionops.infer_conv ~pb:CUMUL env sigma ty2 ty1 with
+      | None -> None
```
Comment (ppedrot) [inline plugins/ltac/rewrite.ml:747]: Don't overindent match branches, I find this style annoying.

Comment (SkySkimmer) [reply plugins/ltac/rewrite.ml:747]: I just do what ocp-indent says.

Comment (Alizter) [reply plugins/ltac/rewrite.ml:747]: https://github.com/OCamlPro/ocp-indent/issues/125

Comment (ppedrot) [reply plugins/ltac/rewrite.ml:747]: There are configuration options, and this default is not sensible. Please stop using the natural law that falls from the sky rhetoric, somebody set those default options and we as users can chose to change them.

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
Merging soon. And just to check, the CI is unrelated right?

**Comment by SkySkimmer** [comment]
CI is unrelated

**Comment by Alizter** [comment]
@coqbot merge now

**Comment by Alizter** [comment]
@coqbot merge now

**Pull request merged**